tstest/natlab/natlab.go

```diff
@@ -266,8 +273,37 @@ func (v PacketVerdict) String() string {
 	}
 }
 
-// A PacketHandler is a function that can process packets.
-type PacketHandler func(p *Packet, inIf *Interface) PacketVerdict
+// A PacketHandler can look at packets arriving at, departing, and
+// transiting a Machine, and filter or mutate them.
+type PacketHandler interface {
+	// HandleIn is invoked when a Packet arrives at a Machine and the
+	// destination IP is an IP owned by the Machine (i.e. it appears
+	// this packet should be delivered to a local socket).
+	//
+	// If HandleIn returns nil, the packet is dropped. HandleIn may
+	// also return an altered Packet, in which case packet processing
+	// on the machine will restart from scratch, and other
+	// PacketHandler methods may be invoked.
+	HandleIn(p *Packet, iif *Interface) *Packet
```

Comment:
the `Packet` type is transparent with several exported fields. So we need to document what the rules are about mutating `p` here. Or we need to unexport the fields of Packet so it's not possible to mutate it.

(So frustrated at GitHub right now for making me type this twice and probably poorly this time.)

Reply:
So, I designed this as a pipeline: function receives a Packet, with which it can do whatever it wants (including mutate its public fields). Natlab is not concurrently touching this Packet, from the moment of the call it's forgotten this Packet exists (although for good measure I also Clone() it when invoking handlers).

The handler is expected to return the packet that should continue to be processed by natlab. It can return the provided Packet verbatim ("firewall allow"), it can return nil ("I'm dropping the packet you gave me, no more work for you to do"), or it an return a completely different Packet ("That packet you gave me needed mangling, here's the fixed one").

The exported fields are therefore by design. You can either return a new `&Packet{...}`, or you can just mutate the one you were given in-place before returning it.

I've tried to alter the docs to better express this. PTAL?

Reply:
(done by video chat; submit when done/happy)